### DISCUSSION TITLE
Back And Forth Implementation

### DISCUSSION THREAD

**Discussion by maheshrajrp** [eclipse-theia/theia#10582] [Q&A] [answered]
Hi,
I'm trying to implement back and forth history in Theia's tree component. Does theia provide with any native history support for `TreeNodes`. It would be great if you could provide some suggestions regarding this. 

Thanks.

**Comment by vince-fugnitto** [comment, marked as answer]
@maheshrajrp sorry for the late reply, the majority of the community was on holiday vacations, as far as I am aware there is no direct native history support for `TreeNodes`. 

Can you elaborate the use-case of why history would need to be maintained? (I could not think of a use-case where the nodes would directly need their history preserved) You can always try to do the caching yourself and re-render the tree based on your state if you require going back and forth.

**Reply by maheshrajrp** [reply]
Hi @vince-fugnitto ,
Not  a problem. Thanks for your reply. It is similar to Editor's Back And Forth. Although, In our case, CustomEditor's option changes iin the  CustomTree based on user's action. As these affect the editor props, maintaining and updating at CustomTree made more sense. However, I was able to create a new Service and was able to implement the same. 

Thanks Again !!
